### PR TITLE
Meta pcengines xen

### PR DESCRIPTION
**Pull request by Asiderr**

### CODE REVIEW

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  - PC engines
+categories:
+  - OS Dev
+  - Security
+  - Manufacturing
+
+---
+# What is the Xen Project?
+
+Xen Project creates a software system that allows the execution of multiple virtual guest operating systems simultaneously on a single physical machine. It is valuable when you want to create a complex security software system. None of the guest operating systems have direct access to the host hardware.
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: @Asiderr  I feel like the term `hypervisor` should be used somewhere in this description.

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+A Xen hypervisor manages the virtual machine guests known as domains. It is controlled by a special domain called domain0 (dom0). Only the hypervisor has access to the physical outputs of the platform. Depending on the guest type (HVM or PV) hypervisor allows unprivileged domains (domUs) to use virtual hardware devices or passes the links to the physical hardware.
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: @Asiderr Please wrap lines at 80chars. Whole paragraph in a single line is not easy to review.

**Comment by Asiderr** [comment]
The asciicast has been archived, in a next commit I will fix it.

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  - OS Dev
+  - Security
+  - Manufacturing
+
+---
+# What is the Xen Project?
+
+Xen Project creates a software system that allows the execution of multiple virtual guest operating systems simultaneously on a single physical machine.
+It is valuable when you want to create a complex security software system.
+None of the guest operating systems have direct access to the host hardware.
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: has

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    <h2><p align=center>Xen Project architecture </p>
+</i>
+
+![Diagram](https://cloud.3mdeb.com/index.php/s/HNmc8yqbcCQzCje/preview)
+
+# How to build the dom0 on the PC Engines APU2?
+
+For this purpose, we will use the [meta-pcengines repository](https://github.com/pcengines/meta-pcengines.git).
+This is the yocto layer, which allows creating a minimal image of the dom0.
+The meta-pcengines needs a kas tool to construct the dom0 build.
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: @Asiderr `kas` could be a hyperlink directing to kas github or docs

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+<i>
+    <h2><p align=center>Xen Project architecture </p>
+</i>
+
+![Diagram](https://cloud.3mdeb.com/index.php/s/HNmc8yqbcCQzCje/preview)
+
+# How to build the dom0 on the PC Engines APU2?
+
+For this purpose, we will use the [meta-pcengines repository](https://github.com/pcengines/meta-pcengines.git).
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: You should mention which commit was used for  this blog post. The current master branch will look differently.

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+The meta-pcengines needs a kas tool to construct the dom0 build.
+Installation and configuration of that tool were described in detail in the
+previous [blog](https://blog.3mdeb.com/2019/2019-02-07-kas/).
+
+Once you have got the kas, clone the [meta-pcengines repository](https://github.com/pcengines/meta-pcengines.git).
+In the main directory, you can see the config directory, the kas configuration,
+and the recipes. The config files of the layer are responsible for setting up
+the platform and distro configuration. The `conf/layer.conf` passes on the
+information where to find the recipes. The kas configuration file contains
+details, how to create the build and. It also inform about repositories that the
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: and. ?

Comment (macpijan) [reply blog/content/post/2019-10-24-meta-pcengines-xen.md]: informs

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+should see a similar log:
+
+```
+NOTE: Executing Tasks
+NOTE: Setscene tasks completed
+NOTE: Tasks Summary: Attempted 2871 tasks of which 0 didn't need to be rerun and all
+ succeeded.
+```
+
+At that point you have to flash a bootable drive. On this spot  I will use the
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: `On this spot` sounds strange to me

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+dd command. Check a directory of the drive to be flashed:
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: drive directory? Rather a device file / device path or something

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+```
+sudo fdisk -l
+```
+
+Change the directory to `<build-dir>/build/tmp/deploy/images/pcengines-apu2` and
+flash the drive:
+
+```
+sudo dd if=xen-dom0-image-pcengines-apu2.hddimg of=/dev/<drive_dir>
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: this is not a directory

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+```
+
+If everything went well, you should see the bootlog:
+
+[![asciicast](https://asciinema.org/a/mzfBdiDWNiWPOEEkev2F6CxwX.svg)](https://asciinema.org/a/mzfBdiDWNiWPOEEkev2F6CxwX?t=95)
+
+# VM guest configuration
+
+Once you have created the dom0 on your APU platform, it is time to place a VM
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: Once you use `apu2`, now `APU`. To be fair, I'm not sure which should be used. Let's verify and be consistent about that.

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+type = "hvm"
+vcpus = 2
+memory = 256
+nographics = 1
+serial = "pty"
+vga="none"
+disk=[ '/mnt/xen-ndvm-image-1.hddimg,,sda,rw' ]
+```
+
+It is required to name your new guest domain and define the boot image
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: If you mean about `/mnt/xen-ndvm-image-1.hddimg,` - t is rather a path to the disk image file, not a directory of any kind

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+directory. A detail description of each variable is given on the [xenbits website](https://xenbits.xen.org/docs/unstable/man/xl.cfg.5.html).
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: detailed

File: blog/content/post/2019-10-24-meta-pcengines-xen.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+```
+
+If the dom0 works well, you should see the boot log of domU:
+
+[![asciicast](https://asciinema.org/a/U17JbsCqhTHdT30tjTg9gjrPf.svg)](https://asciinema.org/a/U17JbsCqhTHdT30tjTg9gjrPf?t=174)
+
+## Conclusions
+
+In answer to the question asked in the title, I will point out that
+meta-pcengines layer represents a simple, quick to use base for xen builds. It
```
Comment (macpijan) [inline blog/content/post/2019-10-24-meta-pcengines-xen.md]: Main goal of meta-pcengines is to enable Yocto builds for pcengines (APU2) boards. Xen is just one of the usecase preseted in this post.

**Comment by macpijan** [comment]
@Asiderr One more thing. Commit message could be shorter (maybe don't include full path). They also should be signed (see our GPG guide).

File: blog/content/post/2019-10-24-meta-pcengines-xen.md (hunk 10 above)
Comment (Asiderr) [reply blog/content/post/2019-10-24-meta-pcengines-xen.md]: I think I should stick to apu2, as it is referred in the [docs](https://pcengines.ch/apu2.htm).

File: blog/content/post/2020-02-05-meta-pcengines-xen.md [hunk 14]
```diff
@@ -0,0 +1,139 @@
+---
+title: 'XEN on the PC Engines apu2 - why you want to use it'
+abstract: 'Xen Project creates a software system that allows the execution of 
+multiple virtual guest operating systems simultaneously on a single physical machine.
```
Comment (artur-rs) [inline blog/content/post/2020-02-05-meta-pcengines-xen.md]: @Asiderr  new lines in the abstract should start with indentation, e.g. 2 spaces or tab (required for correct hugo conversion)

File: blog/content/post/2020-02-05-meta-pcengines-xen.md [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+In this case, it is a PC Engines apu2 platform.'
+cover: /covers/3mdeb-logo_simple-color.png
+author: norbert.kaminski
+layout: post
+published: false
+date: 2020-02-05
```
Comment (artur-rs) [inline blog/content/post/2020-02-05-meta-pcengines-xen.md]: @Asiderr if you want to do local preview, change date to actual or older (just temporarily) and set `published` to `true`

File: blog/content/post/2020-02-05-meta-pcengines-xen.md (hunk 14 above)
Comment (Asiderr) [reply blog/content/post/2020-02-05-meta-pcengines-xen.md]: @artur-rs I've added the indentation

File: blog/content/post/2020-02-05-meta-pcengines-xen.md [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+The repository contains the yocto layer, that allows you to create a
+minimal image of the dom0. To build a usable version of the dom0 image, the
+meta-pcengines needs a [kas tool](https://kas.readthedocs.io/en/0.19.0/index.html).
+Installation, configuration, and some more information about that tool
+you can find in the previous [blog](https://blog.3mdeb.com/2019/2019-02-07-kas/).
+
+Once you have installed the kas, clone the
+[meta-pcengines repository](https://github.com/3mdeb/meta-pcengines/tree/c4ee98ab390b073807173584107c09f49ac1e390).
+To reproduce my results, check if the source commit matches shown above.
+The source consists of configuration files, a kas script, and recipe files.
```
Comment (macpijan) [inline blog/content/post/2020-02-05-meta-pcengines-xen.md]:
```suggestion
The source consists of configuration files, a kas configuration file, and recipe files.
```

File: blog/content/post/2020-02-05-meta-pcengines-xen.md [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+</i>
+
+![Diagram](https://cloud.3mdeb.com/index.php/s/HNmc8yqbcCQzCje/preview)
+
+As shown in the diagram, Domain0 (Dom0) is the privileged domain and it manages
+unprivileged domains (domUs). Dom0 is started by the hypervisor on boot and
+ensures its usability. Therefore it is the Domain that has direct access to the
+hardware. The Domain0 provides hardware connections to DomUs. Dom0 allows full
+virtualized guests to use the virtual hardware and passes the physical links
+paravirtualized domains.
```
Comment (macpijan) [inline blog/content/post/2020-02-05-meta-pcengines-xen.md]: I feel like something is missing in this sentence

File: blog/content/post/2020-02-05-meta-pcengines-xen.md [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+### How to build Dom0 on the PC Engines apu2?
+
+To build the special domain for the PC Engines apu2 board, We will use the
+[meta-pcengines repository](https://github.com/3mdeb/meta-pcengines/tree/c4ee98ab390b073807173584107c09f49ac1e390).
+The repository contains the yocto layer, that allows you to create a
+minimal image of the dom0. To build a usable version of the dom0 image, the
+meta-pcengines needs a [kas tool](https://kas.readthedocs.io/en/0.19.0/index.html).
```
Comment (macpijan) [inline blog/content/post/2020-02-05-meta-pcengines-xen.md]: @Asiderr There were already 1.0 and 2.0 kas releases - please use more recent link to docs. Docs is in 1.0 latest I think: https://kas.readthedocs.io/en/1.0/

Comment (macpijan) [reply blog/content/post/2020-02-05-meta-pcengines-xen.md]: Also, not necessary **needs** - we can as well built it without the kas. Kas is used to manage the yocto layer and provides the docker container for isolation of the development environment.

File: blog/content/post/2020-02-05-meta-pcengines-xen.md [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+Installation, configuration, and some more information about that tool
+you can find in the previous [blog](https://blog.3mdeb.com/2019/2019-02-07-kas/).
+
+Once you have installed the kas, clone the
+[meta-pcengines repository](https://github.com/3mdeb/meta-pcengines/tree/c4ee98ab390b073807173584107c09f49ac1e390).
+To reproduce my results, check if the source commit matches shown above.
+The source consists of configuration files, a kas script, and recipe files.
+The layer configuration specifies settings for the platform and distro.
+It points out where the recipes are placed. The kas script is used by the kas
+tool to clone and checkout `bitbake` layers. It also allows kas to create a
```
Comment (macpijan) [inline blog/content/post/2020-02-05-meta-pcengines-xen.md]: OpenEmbedded layers or Yocto layers or OpenEmbedded/Yocto layers might be better.

File: blog/content/post/2020-02-05-meta-pcengines-xen.md [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+default `bitbake` variables such as MACHINE, DISTRO, etc.
```
Comment (macpijan) [inline blog/content/post/2020-02-05-meta-pcengines-xen.md]: use BitBake if you are not referring to a command

File: blog/content/post/2020-02-05-meta-pcengines-xen.md [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+meta-pcengines needs a [kas tool](https://kas.readthedocs.io/en/0.19.0/index.html).
+Installation, configuration, and some more information about that tool
+you can find in the previous [blog](https://blog.3mdeb.com/2019/2019-02-07-kas/).
+
+Once you have installed the kas, clone the
+[meta-pcengines repository](https://github.com/3mdeb/meta-pcengines/tree/c4ee98ab390b073807173584107c09f49ac1e390).
+To reproduce my results, check if the source commit matches shown above.
+The source consists of configuration files, a kas script, and recipe files.
+The layer configuration specifies settings for the platform and distro.
+It points out where the recipes are placed. The kas script is used by the kas
```
Comment (macpijan) [inline blog/content/post/2020-02-05-meta-pcengines-xen.md]: kas configuration file

File: blog/content/post/2020-02-05-meta-pcengines-xen.md [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+tool to clone and checkout `bitbake` layers. It also allows kas to create a
+default `bitbake` variables such as MACHINE, DISTRO, etc.
+
+Now it is time to create the build. Stay in the meta-pcengines parent directory
+and enter the command:
+
+```
+SHELL=bash kas-docker --ssh-dir ~/ssh-keys build meta-pcengines/kas.yml
```
Comment (macpijan) [inline blog/content/post/2020-02-05-meta-pcengines-xen.md]: Ther is no reference how to get the kas-docker script. The --ssh-dir parameter is not explained. Also it is not needed there (no private repositiories are fetched). just drop it.

File: blog/content/post/2020-02-05-meta-pcengines-xen.md [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+Once you have created the bootable drive, try to boot your platform.
+The following video shows the correct bootlog:
+
+[![asciicast](https://asciinema.org/a/Tr4hhF9sBKC0C9YO5GkwHUrcJ.svg)](https://asciinema.org/a/Tr4hhF9sBKC0C9YO5GkwHUrcJ?t=16)
+
+### Guest VM configuration
+
+Once you have booted the dom0 on your apu2 platform, it is time to launch a VM
+guest. At first write the config file, where you will set up the guest domain
+optins. Here is the example:
```
Comment (macpijan) [inline blog/content/post/2020-02-05-meta-pcengines-xen.md]: options

Comment (macpijan) [reply blog/content/post/2020-02-05-meta-pcengines-xen.md]: *an* example

File: blog/content/post/2020-02-05-meta-pcengines-xen.md [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+
+The following out shows the boot process:
+
+[![asciicast](https://asciinema.org/a/aQfr4P7HneRxkzIN42iHFX3Sd.svg)](https://asciinema.org/a/aQfr4P7HneRxkzIN42iHFX3Sd?t=10)
+
+### Conclusions
+
+The main goal of the meta-pcengines is to enable Yocto builds for pcengines
+(apu2) boards. It provides a good base for various use cases such as Xen.
+IIn future blogs we will show more application examples.
```
Comment (macpijan) [inline blog/content/post/2020-02-05-meta-pcengines-xen.md]: in

**Pull request merged**